app/src/main/java/com/weatherxm/ui/home/quests/steps/GenericStepActivity.kt

```diff
@@ -0,0 +1,282 @@
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.layout.size
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material3.Button
+import androidx.compose.material3.ButtonDefaults
+import androidx.compose.material3.Icon
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.colorResource
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.font.FontWeight
+import com.weatherxm.ui.components.compose.MediumText
+import com.weatherxm.ui.components.compose.Title
+import com.weatherxm.databinding.ActivityGenericStepBinding
+import androidx.compose.ui.unit.dp
+import com.google.firebase.auth.FirebaseAuth
+import com.weatherxm.R
+import com.weatherxm.ui.Navigator
+import com.weatherxm.ui.common.Contracts.ARG_QUEST_STEP
+import com.weatherxm.ui.common.QuestStep
+import com.weatherxm.ui.common.QuestStepType
+import com.weatherxm.ui.common.ctaButtonTitle
+import com.weatherxm.ui.common.parcelable
+import com.weatherxm.ui.common.stepIcon
+import com.weatherxm.util.hasPermission
+import org.koin.android.ext.android.inject
+import org.koin.androidx.viewmodel.ext.android.viewModel
+import org.koin.core.parameter.parametersOf
+import kotlin.getValue
+
+class GenericStepActivity: BaseActivity() {
+    private val model: GenericStepViewModel by viewModel() {
+        parametersOf(intent.parcelable<QuestStep>(ARG_QUEST_STEP))
+    }
+    private val firebaseAuth: FirebaseAuth by inject()
+
+    private lateinit var binding: ActivityGenericStepBinding
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityGenericStepBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        binding.toolbar.setNavigationOnClickListener {
+            onBackPressedDispatcher.onBackPressed()
+        }
+
+        binding.composeView.setContent {
+            Content(
+                model.questStep,
+                onCtaClick = { handleCtaClick() },
+                onSkipClick = { handleSkipClick() }
+            )
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        updateStepState()
+    }
+
+    private fun handleCtaClick() {
+        when (model.questStep.type) {
+            QuestStepType.CONNECT_WALLET -> {
+                // Navigate to wallet connection screen
+            }
+            QuestStepType.ENABLE_LOCATION_PERMISSION -> {
+                // Request location permission
+            }
+            QuestStepType.ENABLE_NOTIFICATIONS -> {
+                // Request notification permission
+                if (!hasPermission(POST_NOTIFICATIONS)) {
+                    navigator.openAppSettings(this)
+                } else {
+                    requestNotificationsPermissions()
+                }
```

Comment:
If we have the permission already, then I guess that `else` branch should call `requestNotificationsPermissions()` but rather call `updateStepState()`, so we immediately terminate successfully.